src/childinfo.c

```diff
@@ -64,72 +58,54 @@ void closeChildInfoPipe(void) {
     }
 }
 
-/* Send COW data to parent. */
-void sendChildInfo(int process_type, int on_exit, size_t cow_size) {
-    if (server.child_info_pipe[1] == -1) return;
-
-    child_info_data buffer = {.process_type = process_type, .on_exit = on_exit, .cow_size = cow_size};
-    ssize_t wlen = sizeof(buffer);
-
-    if (write(server.child_info_pipe[1],&buffer,wlen) != wlen) {
-        /* Nothing to do on error, this will be detected by the other side. */
-    }
-}
-
-/* Update COW data. */
-void updateChildInfo(int process_type, int on_exit, size_t cow_size) {
-    if (!on_exit) {
-        server.stat_current_cow_bytes = cow_size;
-        return;
-    }
-
-    if (process_type == CHILD_TYPE_RDB) {
-        server.stat_rdb_cow_bytes = cow_size;
-    } else if (process_type == CHILD_TYPE_AOF) {
-        server.stat_aof_cow_bytes = cow_size;
-    } else if (process_type == CHILD_TYPE_MODULE) {
-        server.stat_module_cow_bytes = cow_size;
+/* Update Child info. */
+void updateChildInfo(child_info_data *data) {
+    if (data->information_type == CHILD_INFO_TYPE_CURRENT_INFO) {
+        server.stat_current_cow_bytes = data->cow;
+        server.stat_current_save_keys_processed = data->keys;
+        if (data->progress != -1) server.stat_module_progress = data->progress;
+    } else if (data->information_type == CHILD_INFO_TYPE_AOF_COW_SIZE) {
+        server.stat_aof_cow_bytes = data->cow;
+    } else if (data->information_type == CHILD_INFO_TYPE_RDB_COW_SIZE) {
+        server.stat_rdb_cow_bytes = data->cow;
+    } else if (data->information_type == CHILD_INFO_TYPE_MODULE_COW_SIZE) {
+        server.stat_module_cow_bytes = data->cow;
     }
 }
 
-/* Read COW info data from the pipe.
- * if complete data read into the buffer, process type, copy-on-write type and copy-on-write size
- * are stored into *process_type, *on_exit and *cow_size respectively and returns 1.
+/* Read child info data from the pipe.
+ * if complete data read into the buffer, 
+ * data is stored into *buffer, and returns 1.
  * otherwise, the partial data is left in the buffer, waiting for the next read, and returns 0. */
-int readChildInfo(int *process_type, int *on_exit, size_t *cow_size) {
+int readChildInfo(child_info_data *buffer) {
     /* We are using here a static buffer in combination with the server.child_info_nread to handle short reads */
-    static child_info_data buffer;
```

Comment:
I think it made sense when we had only 1 parameter of `cow_size`.
`child_info_data` is static again.

Reply:
Would anyone bother to create a struct in order to avoid passing 4 arguments to a function? I don't think so... 
If we didn't need that struct for the pipe serialized format we wouldn't have had this struct. 

Anyway, with your last commit the comment about the short read and the declaration of the static var is separated. 
I think that the static var and the comment must be in the function that does the `read`ing.  And like before, **only** when it returns success it copies the contents of that struct to the output argument. 
Otherwise the caller may get a partially filled struct.

Reply:
Changed back to original design.

Reply:
Thanks. Sorry for trolling you, I realized there one more thing I'd like to request. 
In the original PR that added it, I asked Yaacov to put the struct definition in childinfo.c, claiming this its layout is private implementation detail of that subsystem and its pipe, and no other c file needs to be aware of it.
I would prefer to move it back there, along with sendChildInfoGeneric that uses it.

Reply:
NP, Done, Thanks for your patience.